mpi-proxy-split/mpi-wrappers/mpi_comm_wrappers.cpp

```diff
@@ -122,6 +178,28 @@ MPI_Comm_free_internal(MPI_Comm *comm)
 
 USER_DEFINED_WRAPPER(int, Comm_free, (MPI_Comm *) comm)
 {
+  // This bit of code is to execute the delete callback function when
+  // MPI_Comm_free is called. Typically we call this function for each
+  // attribute, but since our structure here is a bit different, we check, for
+  // each key value, if a communicator/attribute value pairing exists, and if
+  // it does then we call the callback function.
+  for (auto &tuplePair : tupleMap) {
+    KeyvalTuple *tuple = &tuplePair.second;
+    std::unordered_map<MPI_Comm, void*> *attributeMap = &tuple->_attributeMap;
+    auto attributeMapIter = attributeMap->begin();
+    while (attributeMapIter != attributeMap->end()) {
+      MPI_Comm currComm = attributeMapIter->first;
+      if (currComm == *comm && tuple->_deleteFn != MPI_COMM_NULL_DELETE_FN) {
+        tuple->_deleteFn(*comm,
+                         tuplePair.first,
+                         attributeMap->at(currComm),
+                         tuple->_extraState);
+        attributeMapIter = attributeMap->erase(attributeMapIter);
+        break;
+      }
```

Comment:
1. Why not use `attributeMap->find(*comm)`?
2. If we found the `*comm` in the map, and the `_deleteFn` is null, should we remove it from the map?

Reply:
1. will update it.
2. yes - in fact, we should remove it even if `_deleteFn` is not null. `_deleteFn` just does some clean-up with its arguments, but doesn't actually delete the communicator from the mapping.

thanks for catching these.